File: 206_ReverseLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
//循环（开空间）
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode dummy = null;
        while (head != null) {
            ListNode temp = new ListNode(head.val);
            temp.next = dummy;
            dummy = temp;
            head = head.next;
        }
        return dummy;
    }
}
//循环（不开空间）
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode dummy = null, cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            cur.next = dummy;
            dummy = cur;
            cur = next;
        }
        return dummy;
    }
}
//递归
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return dummy;
    }
}

File: c++/206_ReverseLinkedList.cpp
//sol1
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) return nullptr;
        auto a = head, b = head->next;
        while (b) {
            auto nxt = b->next;
            b->next = a;
            a = b;
            b = nxt;
        }
        head->next = nullptr;
        return a;
    }
};
//sol2
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) return head;
        auto tail = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return tail;
    }
};